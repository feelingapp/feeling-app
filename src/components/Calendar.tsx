import AutoSizer from "react-virtualized-auto-sizer"
import { useEffect, useRef } from "react"
import { VariableSizeList } from "react-window"
import { getWeeksInMonth } from "date-fns"

import Month from "./CalendarMonth"
import { getDateFromIndex, getIndexFromDate, MAX_DATE } from "../utils/dates"

const remToPixels = (rem: number) =>
  rem * parseFloat(getComputedStyle(document.documentElement).fontSize)

function calculateMonthHeight(index: number): number {
  const marginBottom = 24
  const monthTitle = remToPixels(1.6) + 12
  const weekDayTitle = remToPixels(1.2) + 8

  const date = getDateFromIndex(index)
  const weeks = getWeeksInMonth(date)

  const viewportWidth = Math.min(document.documentElement.clientWidth, 600)
  const dayHeight = (viewportWidth - 8 - 8 - 6 * 4) / 7
  const daysHeight = dayHeight * weeks
  const gridGap = 8 * (weeks + 1)

  return monthTitle + weekDayTitle + marginBottom + daysHeight + gridGap
}

function CalendarList({
  width,
  height,
}: {
  width: number
  height: number
}): JSX.Element {
  const listRef = useRef<VariableSizeList>()

  // Scroll to current month
  useEffect(() => {
    listRef.current?.scrollToItem(getIndexFromDate(new Date()), "smart")
  })

  // Rerender list if window is resized
  useEffect(() => {
    const rerenderCalendar = () => listRef.current?.resetAfterIndex(0)
    window.addEventListener("resize", rerenderCalendar)

    return () => window.removeEventListener("resize", rerenderCalendar)
  }, [listRef.current])

  return (
    <VariableSizeList
      itemCount={getIndexFromDate(MAX_DATE)}
      itemSize={calculateMonthHeight}
      height={height}
      width={width}
      ref={listRef}
      overscanCount={3}
    >
      {({ index, style }) => <Month index={index} key={index} style={style} />}
    </VariableSizeList>
  )
}

export default function Calendar(): JSX.Element {
  return (
    <AutoSizer>
      {({ width, height }) => <CalendarList width={width} height={height} />}
    </AutoSizer>
  )
}
