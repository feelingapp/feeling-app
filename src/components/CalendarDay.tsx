import Link from "next/link"
import { styled } from "linaria/react"
import { format, isToday } from "date-fns"

import { Emotion, getEmotionColor, renderEmoji } from "../utils/emotions"

const Container = styled.a`
  display: grid;
  grid-template-areas: "content";
  justify-items: center;
  align-items: center;
  width: 100%;
`

const Background = styled.div<{ backgroundColor?: string; isToday?: boolean }>`
  grid-area: content;
  border-radius: 12px;
  background-color: ${(props) =>
    props.backgroundColor || "var(--light-grey-color)"};
  width: 100%;
  height: 0;
  padding-bottom: 100%;

  box-shadow: ${(props) =>
    props.isToday ? "0 0 0 2px rgba(0, 0, 0, 0.8) inset" : "initial"};
`

const Content = styled.div`
  grid-area: content;
  display: flex;
  flex-direction: column;
  align-items: center;

  svg {
    width: 50%;
    margin-top: var(--space-xxxs);
    filter: drop-shadow(0 0 4px rgba(0, 0, 0, 0.15));
  }
`

const Date = styled.p`
  font-weight: 600;
  font-size: clamp(10px, 2vw, 16px);
  opacity: 0.75;
`

interface Props {
  date: Date
  emotion?: Emotion
  blank?: boolean
}

export default function CalendarDay(props: Props): JSX.Element {
  // Used as an invisible placeholder when a month starts later in the week
  if (props.blank) {
    return (
      <Container as="div">
        <Background backgroundColor="transparent" />
      </Container>
    )
  }

  // Greyed out day when no feelings have been logged for that day
  if (!props.emotion) {
    return (
      <Container as="div">
        <Background isToday={isToday(props.date)} />
      </Container>
    )
  }

  return (
    <Link
      href="/me/day/[date]"
      as={"/me/day/" + format(props.date, "dd-MM-yyyy")}
      passHref={true}
    >
      <Container aria-label={format(props.date, "Mo MMMM yyyy")}>
        <Background
          backgroundColor={getEmotionColor(props.emotion)}
          isToday={isToday(props.date)}
        />
        {props.emotion && (
          <Content>
            <Date>{props.date.getDate()}</Date>
            {renderEmoji(props.emotion)}
          </Content>
        )}
      </Container>
    </Link>
  )
}
