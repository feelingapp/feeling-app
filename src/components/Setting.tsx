import Link from "next/link"
import { styled } from "linaria/react"

const Container = styled.a``

const Title = styled.div`
  font-size: 1.8rem;
  font-weight: 500;
`

const Subtitle = styled.div`
  color: var(--grey-color);
  font-size: 1.4rem;
`

interface Props {
  title: string
  subtitle?: string
  href?: string
}

export default function Setting(props: Props): JSX.Element {
  const { title, subtitle, href } = props

  return (
    <Container href={href}>
      <Title>{title}</Title>
      {subtitle && <Subtitle>{subtitle}</Subtitle>}
    </Container>
  )
}
