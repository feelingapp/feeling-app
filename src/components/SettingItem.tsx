import { styled } from "linaria/react"

const Container = styled.a``

const Title = styled.div`
  font-size: 1.8rem;
  font-weight: 500;
  margin-bottom: var(--space-xxxs);
`

const Subtitle = styled.div`
  color: var(--grey-color);
  font-size: 1.4rem;
`

interface Props {
  title: string
  subtitle?: string
  href?: string
}

export default function SettingItem(props: Props): JSX.Element {
  const { title, subtitle, href } = props

  return (
    <Container as={href ? "a" : "div"} href={href}>
      <Title>{title}</Title>
      {subtitle && <Subtitle>{subtitle}</Subtitle>}
    </Container>
  )
}
