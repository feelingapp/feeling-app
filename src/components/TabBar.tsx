import Link from "next/link"
import { styled } from "linaria/react"
import { useRouter } from "next/router"

import MeIcon from "../assets/icons/me.svg"
import InsightsIcon from "../assets/icons/insights.svg"
import UpsetEmoji from "../assets/emojis/upset.svg"
import { useLayoutEffect, useRef } from "react"

const Container = styled.nav`
  width: 100%;
  display: grid;
  grid:
    ".     fab .    " 5.6rem
    "tab-1 fab tab-2" min-content
    / 1fr auto 1fr;
  position: fixed;
  bottom: 0;
`

const Background = styled.svg`
  grid-row: 2;
  grid-column: -1 / 1;
  display: block;
  width: 100%;
  height: 6.8rem;
  fill: var(--background-color);
  filter: drop-shadow(0 0 4px rgba(0, 0, 0, 0.1));
`

const Tab = styled.a<{ active: boolean }>`
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  color: ${(props) => (props.active ? "inherit" : "var(--grey-color)")};
  font-weight: ${(props) => (props.active ? 500 : 400)};
  z-index: 1;
  padding: var(--space-xs);

  svg {
    width: 2.4rem;
    height: 2.4rem;
    fill: currentColor;
  }
`

const TabLabel = styled.p`
  margin-top: var(--space-xxxs);
  font-size: 1.4rem;
  line-height: 1.6rem;
  user-select: none;
`

const FloatingActionButton = styled.div`
  grid-column: -1 / 1;
  grid-row: -1 / 1;
  z-index: 1;
  padding: var(--space-s);
  background: var(--yellow-color);
  align-self: start;
  justify-self: center;
  border-radius: 50%;
  box-shadow: 0 0 8px rgba(0, 0, 0, 0.1);

  svg {
    width: 3.2rem;
    height: 3.2rem;
  }
`

export default function TabBar(): JSX.Element {
  const { pathname } = useRouter()
  const tabBarRef = useRef(null)

  useLayoutEffect(() => {
    if (tabBarRef.current) {
      console.log(tabBarRef.current.clientHeight)
      document.documentElement.style.setProperty(
        "--tab-bar-height",
        tabBarRef.current.clientHeight + "px"
      )
    }
  }, [tabBarRef])

  return (
    <Container ref={tabBarRef}>
      <Link href="/me" passHref>
        <Tab active={pathname.startsWith("/me")} style={{ gridArea: "tab-1" }}>
          <MeIcon />
          <TabLabel>Me</TabLabel>
        </Tab>
      </Link>

      <Link href="/insights" passHref>
        <Tab
          active={pathname.startsWith("/insights")}
          style={{ gridArea: "tab-2" }}
        >
          <InsightsIcon />
          <TabLabel>Insights</TabLabel>
        </Tab>
      </Link>

      <FloatingActionButton>
        <UpsetEmoji />
      </FloatingActionButton>

      <Background>
        <defs>
          <mask id="fab-cutout">
            <rect
              x="0"
              y="0"
              style={{ width: "100vw", height: "7.2rem", fill: "white" }}
            />
            <circle cx="50%" cy="-2rem" r="4rem" fill="black" />
          </mask>
        </defs>

        <rect x="0" y="0" width="100%" height="100%" mask="url(#fab-cutout)" />
      </Background>
    </Container>
  )
}
