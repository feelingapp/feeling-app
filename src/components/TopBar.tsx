import Link from "next/link"
import Router from "next/router"
import { styled } from "linaria/react"

import BackIcon from "../assets/icons/back.svg?sprite"

const Container = styled.div`
  height: 9.6rem;
  padding: var(--space-s) var(--space-m);
  display: flex;
  user-select: none;
  position: sticky;
  top: 0;
  z-index: 1;
  background: var(--background-color);
`

const Content = styled.div`
  flex: 1;
  align-self: flex-end;
  display: flex;
`

const Text = styled.div`
  flex: 1;
  align-self: flex-end;
`

const Title = styled.h1`
  grid-area: title;
  font-size: 3.2rem;
  font-weight: bold;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`

const Subtitle = styled.h2`
  grid-area: subtitle;
  margin-top: var(--space-xxxs);
  font-size: 1.8rem;
  font-weight: bold;
`

const BackIconLink = styled.a`
  align-self: center;
  margin-right: var(--space-s);
  width: 32px;
  height: 32px;
  cursor: pointer;
`

const IconLink = styled.a`
  grid-area: icon;
  align-self: center;
  width: 32px;
  height: 32px;
`

interface Props {
  title: string
  subtitle?: string
  backButton?: boolean
  rightIcon?: {
    icon: JSX.Element
    href: string
    label: string
  }
}

export default function TopBar(props: Props): JSX.Element {
  const { title, subtitle, backButton, rightIcon } = props

  return (
    <Container>
      <Content>
        {backButton && (
          <BackIconLink
            onClick={() => Router.back()}
            data-test="top-bar-back-icon"
          >
            <BackIcon />
          </BackIconLink>
        )}

        <Text>
          <Title>{title}</Title>
          {subtitle && <Subtitle>{subtitle}</Subtitle>}
        </Text>

        {rightIcon && (
          <Link href={rightIcon.href} passHref>
            <IconLink aria-label={rightIcon.label}>{rightIcon.icon}</IconLink>
          </Link>
        )}
      </Content>
    </Container>
  )
}
