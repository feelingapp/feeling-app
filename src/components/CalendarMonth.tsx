import { CSSProperties } from "react"
import { styled } from "linaria/react"
import { addDays, format, getDaysInMonth, getISODay } from "date-fns/esm"

import CalendarDay from "./CalendarDay"
import { getDateFromIndex } from "../utils/dates"

const Container = styled.div`
  display: flex;
  justify-content: center;

  &:last-child {
    :after {
      content: "";
      display: block;
      position: absolute;
      right: 0;
      bottom: calc(-1 * var(--tab-bar-height));
      height: var(--tab-bar-height);
      width: 100%;
    }
  }
`

const Content = styled.div`
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  gap: clamp(4px, 100vw, 8px);
  margin-bottom: var(--space-m);
  justify-items: center;
  user-select: none;
  max-width: 600px;
  min-width: 270px;
  width: 100%;
`

const MonthTitle = styled.p`
  grid-column: -1 / 1;
  justify-self: center;
  margin-bottom: var(--space-xs);
  color: var(--grey-color);
  font-size: 1.6rem;
  line-height: 1;
`

const WeekDayTitle = styled.div`
  margin-bottom: var(--space-xxs);
  color: var(--grey-color);
  font-size: 1.2rem;
  text-align: center;
  line-height: 1;
`

interface Props {
  index: number
  style: CSSProperties
}

export default function Month(props: Props): JSX.Element {
  const date = getDateFromIndex(props.index)
  const numberOfDays = getDaysInMonth(date)
  const day = getISODay(date) % 7

  return (
    <Container style={props.style}>
      <Content>
        <MonthTitle>{format(date, "MMMM yyyy")}</MonthTitle>

        <WeekDayTitle>M</WeekDayTitle>
        <WeekDayTitle>T</WeekDayTitle>
        <WeekDayTitle>W</WeekDayTitle>
        <WeekDayTitle>T</WeekDayTitle>
        <WeekDayTitle>F</WeekDayTitle>
        <WeekDayTitle>S</WeekDayTitle>
        <WeekDayTitle>S</WeekDayTitle>

        {/* Add blank calendar days so actual ones are inline with week day titles */}
        {Array.from({ length: day }).map((_, index) => (
          <CalendarDay date={date} blank={true} key={index} />
        ))}

        {Array.from({ length: numberOfDays }).map((_, index) => (
          <CalendarDay date={addDays(date, index)} key={index} />
        ))}
      </Content>
    </Container>
  )
}
