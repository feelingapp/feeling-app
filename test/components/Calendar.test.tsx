import { shallow } from "enzyme"
import { advanceTo } from "jest-date-mock"

import * as Calendar from "../../src/components/Calendar"

describe("Calendar: ", () => {
  it("matches snapshot", () => {
    advanceTo(new Date(2021, 0, 1))

    const wrapper = shallow(<Calendar.default />)
    expect(wrapper).toMatchSnapshot()
  })

  it("calendar list matches snapshot", () => {
    const wrapper = shallow(<Calendar.CalendarList width={400} height={800} />)
    expect(wrapper).toMatchSnapshot()
  })

  it("converts rem to pixels", () => {
    document.documentElement.style.setProperty("font-size", "10px")
    const result = Calendar.remToPixels(1.6)
    expect(result).toEqual(16)
  })

  it("calculates month height", () => {
    document.documentElement.style.setProperty("font-size", "10px")

    jest
      .spyOn(document.documentElement, "clientWidth", "get")
      .mockImplementation(() => 1000)

    const fiveRowMonth = Calendar.calculateMonthHeight(0)
    expect(fiveRowMonth).toEqual(520)

    const sixRowMonth = Calendar.calculateMonthHeight(4)
    expect(sixRowMonth).toEqual(608)
  })
})
