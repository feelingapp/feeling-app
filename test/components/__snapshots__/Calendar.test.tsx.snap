// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar:  calendar list matches snapshot 1`] = `
<List
  direction="ltr"
  height={800}
  itemCount={20}
  itemSize={[Function]}
  layout="vertical"
  overscanCount={3}
  useIsScrolling={false}
  width={400}
>
  <Component />
</List>
`;

exports[`Calendar:  matches snapshot 1`] = `
<AutoSizer
  disableHeight={false}
  disableWidth={false}
  onResize={[Function]}
  style={Object {}}
>
  <Component />
</AutoSizer>
`;
